### PR TITLE
refactor

### PR DESCRIPTION
**Pull request by JasonZhongZexin**
[SVF_redis-server.txt](https://github.com/SVF-tools/SVF/files/9209667/SVF_redis-server.txt)
[SVF-1_redis-server.txt](https://github.com/SVF-tools/SVF/files/9209668/SVF-1_redis-server.txt)

### CODE REVIEW

**Comment by mbarbar** [comment]
Hey, apologies I couldn't get to this last night, but will tonight.

File: include/MemoryModel/SymbolTableInfo.h [hunk 1]
```diff
@@ -165,6 +163,16 @@ class SymbolTableInfo
     // @{
     static bool isNullPtrSym(const Value *val);
     static bool isBlackholeSym(const Value *val);
+    static bool argInNoCallerFunction(const Value *val);
```
Comment (mbarbar) [inline include/MemoryModel/SymbolTableInfo.h:166]: Does this mean val is an argument of an uncalled function?

If so, maybe `argOfUncalledFunction` may be clearer.

File: lib/SVF-FE/LLVMUtil.cpp [hunk 2]
```diff
@@ -91,17 +91,22 @@ bool LLVMUtil::isConstantObjSym(const Value *val)
 /*!
  * Return reachable bbs from function entry
  */
-void LLVMUtil::getFunReachableBBs (const Function * fun, DominatorTree* dt, std::vector<const BasicBlock*> &reachableBBs)
+void LLVMUtil::getFunReachableBBs (const SVFFunction* svffun, std::vector<const BasicBlock*> &reachableBBs)
```
Comment (mbarbar) [inline lib/SVF-FE/LLVMUtil.cpp:94]: `svfFun` (I'd personally prefer `fun` even)

File: lib/SVF-FE/LLVMUtil.cpp [hunk 3]
```diff
@@ -91,17 +91,22 @@ bool LLVMUtil::isConstantObjSym(const Value *val)
...
 {
+    assert(!SVFUtil::isExtCall(svffun) && "The calling function cannot be an external function.");
+    //initial DominatorTree
+    DominatorTree dt;
+    dt.recalculate(*svffun->getLLVMFun());
```
Comment (mbarbar) [inline lib/SVF-FE/LLVMUtil.cpp:99]: Is `getFunReachableBBs` called more than once per function? If so, I think `dt.recalculate` on every call is not the best idea.

File: lib/SVF-FE/SymbolTableBuilder.cpp [hunk 4]
```diff
@@ -205,9 +205,76 @@ void SymbolTableBuilder::buildMemModel(SVFModule* svfModule)
...
+        else if (LLVMUtil::functionDoesNotRet(fun))
+        {
+            symInfo->getModule()->addFunctionDoesNotRet(fun);
+        }
+        if (!isExtCall(svfFun))
+        {
+            const BasicBlock* exitBB = LLVMUtil::getFunExitBB(fun);
+            symInfo->getModule()->addFunExitBB(fun,exitBB);
+
+            for (Function::const_iterator bit = fun->begin(), ebit = fun->end(); bit != ebit; ++bit)
```
Comment (mbarbar) [inline lib/SVF-FE/SymbolTableBuilder.cpp:235]: I don't fully understand this part, I'll have to leave to @yuleisui

File: lib/SVF-FE/SymbolTableBuilder.cpp [hunk 5]
```diff
@@ -205,9 +205,76 @@ void SymbolTableBuilder::buildMemModel(SVFModule* svfModule)
...
-        symInfo->blackholeSyms.insert(val);
+        symInfo->getModule()->addBlackholeSyms(val);
+}
+
+
+void SymbolTableBuilder::collectSpecialSym(const Value* val){
+    if (LLVMUtil::isPtrInDeadFunction(val))
+    {
+        symInfo->getModule()->addPtrInDeadFunction(val);
+    }
```
Comment (mbarbar) [inline lib/SVF-FE/SymbolTableBuilder.cpp:218]: An empty line here will make it clearer it is a different if tree

**Review by mbarbar — CHANGES_REQUESTED** [review]
Most important changes are returning references and double `find`s. They do actually affect performance.

Changes should be small, but other than those should be good :)

File: lib/SVF-FE/SymbolTableBuilder.cpp (hunk 4 above)
Comment (yuleisui) [reply lib/SVF-FE/SymbolTableBuilder.cpp:236]: add fields (DeadFunction, functionDoesNotRet and FunExitBB) in SVFFunction for these three attributes

File: include/Util/SVFModule.h [hunk 6]
```diff
@@ -209,6 +221,144 @@ class SVFModule
...
+    inline void addBlackholeSyms(const Value *val){
+        blackholeSyms.insert(val);
+    }
+
+    inline void addArgOfUncalledFunction(const Value *val)
+    {
+        argOfUncalledFunction.insert(val);
+    }
+
+    inline void addDeadFunction(const Function *fun)
```
Comment (yuleisui) [inline include/Util/SVFModule.h:353]: isUncalledFunction

File: lib/SVF-FE/SymbolTableBuilder.cpp [hunk 7]
```diff
@@ -205,9 +205,77 @@ void SymbolTableBuilder::buildMemModel(SVFModule* svfModule)
...
+    {
+        if (LLVMUtil::isReturn(inst))
+            symInfo->getModule()->addReturn(inst);
+    }
+    else if (const PointerType * ptrType = SVFUtil::dyn_cast<PointerType>(val->getType()))
+    {
+        const Type* type = LLVMUtil::getPtrElementType(ptrType);
+        symInfo->getModule()->addptrElementType(ptrType, type);
+    }
+    else if (LLVMUtil::argOfUncalledFunction(val))
```
Comment (yuleisui) [inline lib/SVF-FE/SymbolTableBuilder.cpp:275]: make sure mutually exclusive.
change to if

File: lib/MemoryModel/SymbolTableInfo.cpp [hunk 8]
```diff
@@ -323,15 +321,105 @@ void SymbolTableInfo::destroy()
...
+
+bool SymbolTableInfo::isReturn(const Instruction *inst)
+{
+    const Set<const Instruction*> isReturnSet = symInfo->getModule()->getIsReturn();
+    return isReturnSet.find(inst) != isReturnSet.end();
+}
+
+bool SymbolTableInfo::functionDoesNotRet(const Function *fun)
+{
+    const Set<const Function*> functionDoesNotRetSet = symInfo->getModule()->getFunctionDoesNotRet();
```
Comment (yuleisui) [inline lib/MemoryModel/SymbolTableInfo.cpp:348]: const Set<const Function*> => const Set<const Function*>&

File: lib/MemoryModel/SymbolTableInfo.cpp [hunk 9]
```diff
@@ -323,15 +321,105 @@ void SymbolTableInfo::destroy()
...
+    return functionDoesNotRetSet.find(fun) != functionDoesNotRetSet.end();
+}
+
+bool SymbolTableInfo::isPtrInDeadFunction (const Value * value)
+{
+    const Set<const Value*> ptrInDeadFunctionSet = symInfo->getModule()->getPtrInDeadFunction();
```
Comment (yuleisui) [inline lib/MemoryModel/SymbolTableInfo.cpp:354]: const Set<const Value*> ptrInDeadFunctionSet => const Set<const Value*>& ptrInDeadFunctionSet

File: include/Util/SVFModule.h [hunk 10]
```diff
@@ -221,61 +221,61 @@ class SVFModule
...
 
-    inline const Set<const Value*> getArgInNoCallerFunction() const
+    inline const Set<const Value*>& getArgOfUncalledFunction() const
     {
-        return argInNoCallerFunction;
+        return argOfUncalledFunction;
     }
 
-    inline const Set<const Function*> getIsDeadFunction() const
+    inline const Set<const Function*>& getIsUncalledFunction() const
```
Comment (mbarbar) [inline include/Util/SVFModule.h]: `getUncalledFunctions` is a better name I think, unless I am misunderstanding

File: include/Util/SVFModule.h [hunk 11]
```diff
@@ -221,61 +221,61 @@ class SVFModule
...
+    inline const Set<const Value*>& getBlackholeSyms() const
     {
         return blackholeSyms;
     }
 
-    inline const Set<const Value*> getArgInNoCallerFunction() const
+    inline const Set<const Value*>& getArgOfUncalledFunction() const
     {
-        return argInNoCallerFunction;
+        return argOfUncalledFunction;
```
Comment (mbarbar) [inline include/Util/SVFModule.h:233]: Maybe `getArgsOfUncalledFunction`?

File: include/Util/SVFModule.h [hunk 12]
```diff
@@ -221,61 +221,61 @@ class SVFModule
...
 
-    inline const Set<const Function*> getIsDeadFunction() const
+    inline const Set<const Function*>& getIsUncalledFunction() const
     {
-        return isDeadFunction;
+        return isUncalledFunction;
     }
 
-    inline const Set<const Instruction*> getIsReturn() const 
+    inline const Set<const Instruction*>& getIsReturn() const
```
Comment (mbarbar) [inline include/Util/SVFModule.h:236]: `getReturns`

File: include/Util/SVFModule.h [hunk 13]
```diff
@@ -221,61 +221,61 @@ class SVFModule
...
     }
 
-    inline const Set<const Instruction*> getIsReturn() const 
+    inline const Set<const Instruction*>& getIsReturn() const 
     {
         return isReturns;
     }
 
-    inline const Set<const Function*> getFunctionDoesNotRet() const
+    inline const Set<const Function*>& getFunctionDoesNotRet() const
```
Comment (mbarbar) [inline include/Util/SVFModule.h]: `getNonReturningFunctions`

File: include/Util/SVFModule.h [hunk 14]
```diff
@@ -343,14 +345,14 @@ class SVFModule
...
     }
 
-    inline void addArgInNoCallerFunction(const Value *val)
+    inline void addArgOfUncalledFunction(const Value *val)
     {
-        argInNoCallerFunction.insert(val);
+        argOfUncalledFunction.insert(val);
     }
 
     inline void addDeadFunction(const Function *fun)
```
Comment (mbarbar) [inline include/Util/SVFModule.h]: Maybe `addUncalledFunction`

File: include/Util/SVFModule.h [hunk 15]
```diff
@@ -221,61 +221,61 @@ class SVFModule
...
     }
 
-    inline const Set<const Function*> getFunctionDoesNotRet() const
+    inline const Set<const Function*>& getFunctionDoesNotRet() const
     {
         return functionDoesNotRet;
     }
 
-    inline const Set<const Value*> getPtrInDeadFunction() const 
+    inline const Set<const Value*>& getPtrInDeadFunction() const
```
Comment (mbarbar) [inline include/Util/SVFModule.h:241]: `getPtrsInUncalledFunctions` or something similar

File: include/Util/BasicTypes.h [hunk 16]
```diff
@@ -130,19 +130,21 @@ class SVFFunction : public SVFValue
     bool isIntri;
     Function* fun;
     std::vector<const BasicBlock*> reachableBBs;
+    mutable bool isUncalled;
+    mutable bool isNotRet;
```
Comment (mbarbar) [inline include/Util/BasicTypes.h:135]: I know that `setIsUncalled` is called below, but is this necessary, as in, is there a way to achieve the same thing without using `mutable`? Can we know these values at construction time. `mutable` is a bit scary haha.

**Review by mbarbar — APPROVED** [review]
Seems to me everything was addressed, but I made a few more small comments.

As long as AvgTopLvlPts is still good and you consider the comments, I'm happy.

**Pull request merged**